pywbem_mock/_baseprovider.py

```diff
@@ -81,12 +81,18 @@ class BaseProvider(object):
     executed directly.
     """
 
-    def __init__(self):
+    def __init__(self, cimrepository=None, provider_registry=None):
```

Comment:
These two variables were already in the BaseProvider superclass before this change. See https://github.com/pywbem/pywbem/blob/master/pywbem_mock/_baseprovider.py#L84.
The code before this change just overwrote them in the derived classes.

I have no problem if an abstract superclass has properties (even public ones), or implemented methods, as long as that is part of what it intends to provide. The only constraint on abstract classes is that they cannot be instantiated, but that is different from not having properties or implemented methods. Not sure whether the various ways to declare abstract classes in Python prevent instantiation of the abstract class, but currently, neither BaseProvider nor the other abstract base classes have been declared abstract at the class level. Maybe that is the way to go, to add the abstract declaration that fits to the abstractmethod/abstractproperty decorators.

Reply:
I thought about making BaseProvider an abstract but to me the role of an abstract class is primarily to define methods to be implemented by subclasses. Since the subclasses of BaseProvider (MainProvider and InstanceWriteProvider have no need of a template for any APIs there is no reason for having Abstract methods.

The other methods in BaseProvider are the support methods for the request responder methods in MainProvider and InstanceWriteProvider, i.e. those methods that would be used in both of them and also extended to user-defined providers. 

Therefore, It is not really a clean candidate for an abstract class.

It is primarily a container for a bunch of methods that are used in subclasses but not modified by those subclasses.

Reply:
We agreed to merge this change and discuss lateron how to refactor it.